### ISSUE TITLE
Improve handling of GWC filter parameter storage

### ISSUE THREAD

**Issue by aaime** [geosolutions-it/eumetsat-EOWS#361]
GWC stores filter parameters used for tiles in a property file. There is one property file for each layer, containing a map between the parameter hash and the actual parameter values.

Issues:

- The file is read and re-written every time a tile with parameters is stored
- Writes are direct to the file
- The property file can grow very large

Short term goals to reduce pressure on these files:

- Re-write the file only if the property was missing (now it's not even checking)
- Maybe keep a cache of the last few recently written properties and compare with the last file modification time to avoid reading it at all
- When writing, write on the side and use an atomic rename

Longer term goals:
- Switch away from property file, allowing usage of a storage that can scale better and work with multiple concurrent processes
- Check what MapCache does to manage the same (and maybe MapProxy too?)
- Create a storage for these properties that can be shared among all blob stores, S3 and Azure are basically using the same property file machinery

**Comment by simboss** [comment]
@aaime as I mentioned during the call, a short term goal IMHO is also to reduce the size of these files on disk as they can easily become many MB each. At the moment we have 20k time bins for each mosaic and that keeps growing.
Throw that in the mix.

**Comment by simboss** [comment]
Random thought, why do we need to keep that properties files around?

I mean if there one way to hash the filter combination and then one way to get to its directory we could do directly no?
Just thinking out loud...

**Comment by simboss** [comment]
I'd suggest to have a look at MapCache approach as well.

**Comment by simboss** [comment]
Notice that right now we are asking GeoServer to cachel all styles even if it is not needed.

**Comment by aaime** [comment]
Checked MapCache a bit:

- It allows dimensions, and multiple dimension
- Valid values for dimensions can come from static lists, regexes, but also a database
- The seeding tool accepts multiple dimensions to be specified when seeding/truncating a cache

Also checked GWC better. The list of parameters is not used when purging a particular parameter combination (for that one, the hash is computed and it drives the removal). The issue is that in GWC (and in particular, GeoServer) a parameter definition can be removed from the layer at runtime... and then GWC has to locate all tile caches that had such parameter, and remove them. This cannot be done using just the hashes.

MapCache does not even consider this problem, it's up to the admin to deal with it.

By the way, it's interesting to check how MapCache does locking.. it's either file system based, or memcached, or both: https://mapserver.org/mapcache/locks.html

**Comment by aaime** [comment]
Found out that parameter storage is also used for another task, the "drop by bbox" which drops tiles by bounds across all of the layer caches. To use the existing machinery, it first collects all the available parameters, and then create a truncate job for the a given parameter value set. Could be rewritten to just use the hash, but it would require changes in multiple places, and multiple blob stores.

Let's go back to the simpler approach or improving property files handling.

**Comment by aaime** [comment]
Plan to improve the handling of property files, overview:

- Make sure we handle the property file only when strictly needed, that is, when the directory for a new set of parameters is created. It's a single time per directory, however, the "metadata.properties" is still unique for the entire layer, so, while this reduces very significantly reads and writes, it does not remove concurrency.
- When writing, make sure the properties was actually modified before rewriting it out
- When writing, use an optimistic locking approach, if the target file was modified while we were prepping its new contents, start over.

Detailed plan:
- Modify [getFileHandleTile](https://github.com/GeoWebCache/geowebcache/blob/03042acb36e2b92ecbb4317626e69f204f1d8cab/geowebcache/core/src/main/java/org/geowebcache/storage/blobstore/file/FileBlobStore.java#L500) to accept a Lambda that is invoked when 
- When [adding a property](https://github.com/GeoWebCache/geowebcache/blob/34a9fca5aeb9d01e744c1779406f1a5010f619ba/geowebcache/core/src/main/java/org/geowebcache/storage/blobstore/file/FileBlobStore.java#L646) to the file, still check if it was already there with he same value, if so, do not write
- When modifying the property file, write to a sidecar file with a random name, and then do a atomic rename. The code should be pretty similar to [writeFile](https://github.com/GeoWebCache/geowebcache/blob/34a9fca5aeb9d01e744c1779406f1a5010f619ba/geowebcache/core/src/main/java/org/geowebcache/storage/blobstore/file/FileBlobStore.java#L531) except that it should be writing a Properties. There is the temptation to just reuse the method, keeping in memory a serialized byte[] out of the property, let's avoid it, we cannot take that kind of memory pressure, the properties file can grow large.
- When modifying the property file, store the last modification date of the file, and right before doing the atomic rename, compare it: if it's not the same, start over, re-read the contents, store the new date, and so on (optimistic locking approach).

An implementation for GWC master, plus a backport, should be doable in 1.5 days, 2 tops.

**Comment by nmco** [comment]
Following @aaime investigation I did a few tests to understand the code and end up having a few questions \ taking a few notes.

The code that deals with the generation of the hash for a group of parameters is available in the [class](https://github.com/GeoWebCache/geowebcache/blob/34a9fca5aeb9d01e744c1779406f1a5010f619ba/geowebcache/core/src/main/java/org/geowebcache/filter/parameters/ParametersUtils.java#L32). The [SHA-1 algorithm](https://github.com/GeoWebCache/geowebcache/blob/34a9fca5aeb9d01e744c1779406f1a5010f619ba/geowebcache/core/src/main/java/org/geowebcache/filter/parameters/ParametersUtils.java#L32) is used to generate the hash.  The path for the tile retrieving or storage is build [here](https://github.com/GeoWebCache/geowebcache/blob/f69d7b507680882c7c3b079aba4b9ea6894e360c/geowebcache/core/src/main/java/org/geowebcache/storage/blobstore/file/DefaultFilePathGenerator.java#L78-L81), **looks like GWC assumes that no collisions will happen for the generated hashes (which is not true)**, the meta-data property file is not used for managing hashes collisions. 

**Looks like that the only propose of the meta-data properties file is too support operations that need to discover which parameters have been created.** @aaime looks like you mention two operations that require the meta-properties, I could only find the bounding box based truncate operation (I will need your assistance to find the second):
-  [Truncate by a bounding box only.](https://github.com/GeoWebCache/geowebcache/blob/5e913193ff50a61ef9dd63a87887189352fa6b21/geowebcache/core/src/main/java/org/geowebcache/seed/TruncateBboxRequest.java#L55)

Is worth mentioning that:
- If a [truncate submitted thought the GWC form](https://github.com/GeoWebCache/geowebcache/blob/a2e5c33a5e892af15bd9501eaf18f5391f2ff4d2/geowebcache/rest/src/main/java/org/geowebcache/rest/service/FormService.java#L175) doesn't rely on the meta-data property file, [it will use the provided parameters values](https://github.com/GeoWebCache/geowebcache/blob/34a9fca5aeb9d01e744c1779406f1a5010f619ba/geowebcache/core/src/main/java/org/geowebcache/storage/TileRange.java#L60).
- If an empty cache is issue for a layer the parameters based caches will be found out by [listing the layer cache directory content](https://github.com/GeoWebCache/geowebcache/blob/34a9fca5aeb9d01e744c1779406f1a5010f619ba/geowebcache/core/src/main/java/org/geowebcache/storage/blobstore/file/FileBlobStore.java#L306).

**This make me wonder:
- Could we make the meta-data file creation optional, when deactivated the operations that really on it would be deactivated too.
- Could we make the operations that depend on the reverse parameter mapping actually obtain the parameters from the file system? Like it's done for [the layer full truncate](https://github.com/GeoWebCache/geowebcache/blob/34a9fca5aeb9d01e744c1779406f1a5010f619ba/geowebcache/core/src/main/java/org/geowebcache/storage/blobstore/file/FileBlobStore.java#L306)?**

Notes and references:
1. Looks like this reverse parameters mapping was added by Kevin in 2017 with his [commit](https://github.com/GeoWebCache/geowebcache/commit/3d90e12da96a72e2178c22fe723474b4ca9ff9e5).
1. When a miss happens, the tile are wrote to the disk (blob store) first before replying, should be the opposite.
1. Where to configure new parameters (e.g. time dimension) is not documented anywhere.
1. The issue described here is particularity bad when a time dimension (or other parameter with a lot of possible values its used), the meta-data file will keep growing bigger and bigger:
```
#auto generated file, do not edit by hand
#Wed Sep 02 20:51:36 WEST 2020
parameters.147beae8d38d7edb5c4658fc8013d91eb6d308ce=STYLES%3Dpolygon_red%26TIME%3D2020-09-01T00%253A00%253A00.0Z%252F2020-09-03T00%253A00%253A00.0Z
parameters.7aa18232c29f121c15fe43747a4d4b3795353927=STYLES%3Dpolygon%26TIME%3D2020-09-02T00%253A00%253A00.0Z%252F2020-09-04T00%253A00%253A00.0Z
parameters.27a323b1ce581baa1246ced9e6dba0942fd0f068=STYLES%3Dpolygon_green
parameters.c78e2b130f5baa8bf1e04514b33315ad3cb8e2c7=STYLES%3Dpolygon_red%26TIME%3D2020-09-02T00%253A00%253A00.0Z%252F2020-09-04T00%253A00%253A00.0Z
parameters.dfe0445f28c40fe1ca15f193e88e9b9845aa1a8d=STYLES%3Dpolygon_green%26TIME%3D2020-09-02T00%253A00%253A00.0Z%252F2020-09-04T00%253A00%253A00.0Z
```
1. Regarding MapCache, the configuration is explained [here](https://mapserver.org/mapcache/dimensions.html#first-level-dimensions):
  1. The MapCache code that builds the file path for a certain tile is available [here](https://github.com/mapserver/mapcache/blob/master/lib/cache_disk.c#L128-L155), looks like the value of the dimension is used directly.
  1. [Templates](https://mapserver.org/mapcache/dimensions.html#storing-dimensions) can be used to define the path used by the file based store, so yes looks like is up to he admin to configure the a store configuration.
1. I had to time-box my mapcache investigation, I would need to spend a bit more of time to understand how it works.

**Comment by aaime** [comment]
Known usages of parameters stored in property files:
* Removal by bounding box, across all parameter related caches (could be rewritten to just go in each cache directory, not fully trivial)
* Removal or re-config of a parameter filter, to locate and delete all caches that are out of date (e.g., say one removes a valid value from the list of parameter valid values, seems the code would go, locate the right directories, and remove them)
* Removal of orphans, that is, directories that do not match anymore the configuration of the layer. There is a dedicated GWC REST call that can be used to initiate that (I'm guessing, to clean up an older cache that was not using parameters, or that fell out of synch for some reason)

**Comment by nmco** [comment]
Summarizing the decisions that have been taken:
- Short term plan, the quick fix of disabling the writing on the meta-data has been implemented on EUMESAT branch.
- Mid-term \ long-term plan is to implemented the fix suggested by @aaime in [this comment](https://github.com/geosolutions-it/eumetsat-EOWS/issues/361#issuecomment-684725715) this fix has two main tasks:
 - Only write and access the file when really needed.
 - Implement an optimist locking mechnaism to access the file.

**Comment by nmco** [comment]
@jemacchi as discussed let's start by:
- Reproducing the problem locally, in this case reproducing the problem will be quite difficult since you would need quite a load, in this case I would recommend to set-up GS and GWC and check \ validate that the meta-data properties file is created and access as described on this issue.
- Then you should validate the plan proposed by @aaime [here](https://github.com/geosolutions-it/eumetsat-EOWS/issues/361#issuecomment-684725715), you should check if you are comfortable with the estimate and clarify all the technical questions you may have.
- Then let's sync and decide when to move on whit the fix implementation.

**Comment by jemacchi** [comment]
@nmco 
1. I could reproduce problem locally. Debugged general flow/calls stack, familiarizing with GWC code. Thanks for assistance ! (@fernandor777 and you )
2. I'm not sure If -based on my knowledge of code- I can do it in top 2 days as proposed. 
I would say that at least I will need 2 days, 3 days max. 
Note: I'm considering also testing (as you commented) and the backport. Regarding last item, which releases should be considered ? (probably I will ask you for assistance on that).
3. Based on you comment [here](https://github.com/geosolutions-it/eumetsat-EOWS/issues/361#issuecomment-687235986), I think you are tailoring the solution proposed by @aaime [here](https://github.com/geosolutions-it/eumetsat-EOWS/issues/361#issuecomment-684725715). Should I include the gz feature ?

Let me know your thoughts

**Comment by jemacchi** [comment]
Regarding technical questions:

@aaime  from [here](https://github.com/geosolutions-it/eumetsat-EOWS/issues/361#issuecomment-684725715)  when you comment not to delegate to writeFile, which is the reason ? (I see that writeFile is only called from [FileBlobStore.put](https://github.com/GeoWebCache/geowebcache/blob/34a9fca5aeb9d01e744c1779406f1a5010f619ba/geowebcache/core/src/main/java/org/geowebcache/storage/blobstore/file/FileBlobStore.java#L473) method)

**Comment by aaime** [comment]
Because I said to write only on directory creation, and because "writeFile" should "write a file", otherwise the method should be "write a file and shared parameter metadata"

**Comment by jemacchi** [comment]
@aaime, ok. Then, it's just about refactoring. I can rename the method correctly and split responsibilities.

**Comment by nmco** [comment]
Thx for the updates @jemacchi, please see my replies bellow:

> I'm not sure If -based on my knowledge of code- I can do it in top 2 days as proposed.
> I would say that at least I will need 2 days, 3 days max.
> Note: I'm considering also testing (as you commented) and the backport. Regarding last item, which releases should be considered ? (probably I will ask you for assistance on that).

I see, let's move forward, keep me posted on a any blocker or issue that may increase the estimate.

> Based on you comment here, I think you are tailoring the solution proposed by @aaime here. or should I include the gz feature ? or time-box atttributes considerations?

Only the later, [this one](https://github.com/geosolutions-it/eumetsat-EOWS/issues/361#issuecomment-684725715).

**Comment by jemacchi** [comment]
@aaime could you please take a look at this [PR](https://github.com/GeoWebCache/geowebcache/pull/879) ? thanks

**Comment by jemacchi** [comment]
@aaime 
based on latest talk about concurrency test:
1) added system properties to define the MAX_RW_ATTEMPTS
2) In order to catch all the warning messages (from stress test) for each case on which an attempt to write is not working, it's required to update the putLayerMetadata method signature (that is throwing a StorageException). I checked alternative ways to do it without changing signature, but unfortunatelly it seems the only way to achieve it and that would imply to adjust also other classes implementing BlobStore (the abstract class that defines the method). 
Thoughs ? suggestions ?

**Comment by jemacchi** [comment]
@aaime just please avoid about last message, point 2.
Basically, after long time looking at code, I found what was happening. Concurrence check was failing in those cases on which temporary file was beat by another process that already updated the keys. So in that case it was necessary to re-read the properties again and add the new KVP if required.
In the middle of process, and for the sake of code sanity, I moved all LayerMetadata related methods into a separate class (since it was really hard trying to detect what was happening in the middle of previous code).
I have tested manually and it's working. Also I have enabled all the tests (included those that are aggressively triggering threads, and it's working too).
Please take a look at the [PR](https://github.com/GeoWebCache/geowebcache/pull/879) and let me know if it's ok.

CC: @simboss @nmco

**Comment by aaime** [comment]
> Basically, after long time looking at code, I found what was happening. Concurrence check was failing in those cases on which temporary file was beat by another process that already updated the keys. So in that case it was necessary to re-read the properties again and add the new KVP if required.

The timestamp checks were supposed to limit this possibility (while we knew it was still a possibility). Looking at the code, it does not check that the timestamp has changed between writing the temp file, and doing the rename. Writing the temp file is IO, very slow compared to the instructions running on a CPU. Added a note in the comment.
Will it solve the issue? No, it still can happen. But it should reduce the window significantly.

> In the middle of process, and for the sake of code sanity, I moved all LayerMetadata related methods into a separate class (since it was really hard trying to detect what was happening in the middle of previous code).

Good call.

> I have tested manually and it's working.

Manually how? Multiple GWCs running in parallel?

> Also I have enabled all the tests (included those that are aggressively triggering threads, and it's working too).
> Please take a look at the [PR](https://github.com/GeoWebCache/geowebcache/pull/879) and let me know if it's ok.

Yep, it's all green, but I see no tests using multiple blob stores, so the tests are actually covered by the newly introduced RW lock code, which is definitely more robust. In a sense, good, but at the same time, would be good having a test that uses multiple blob stores, at least for local testing (I guess it will eventually fail, every now and then, so it will have to be ignored. Optimistic locking based on last modification date of files only reduces the chance of conflict, does not eliminate it fully... full distributed locking would be needed to handle things properly).

**Comment by jemacchi** [comment]
@nmco @aaime   

I think we are done with the GWC issue (included your comment about checking content and repeating if required)

Concurrent tests based on the same FileBlobStore (pessimistic lock) are working fine.

Tests related to multiple GWC(stores) instances will be marked as @Ignore, since when running eventually could fail (we know that). Besides that, we will have a couple of tests available -with multiple FileBlobStores- and also logging on code that will allow to keep track of those cases on which a read/write action was not completed (reaching max retry attempts)

Here is the [PR](https://github.com/GeoWebCache/geowebcache/pull/879)

**Comment by nmco** [comment]
The [PR](https://github.com/GeoWebCache/geowebcache/pull/879) has been merged on master.

@aaime I have a few questions where I would need your help:
- I guess we need to backport this to stable at least 1.18.x ?
- We need to backport this to EUMETSAT custom branch (I don't know the branch)?
- @jemacchi pleas make sure you update GWC issue accordingly.

**Comment by aaime** [comment]
@nmco:
* I would backport yes. Maybe on 1.17.x too
* Personally I would not touch the EUMETSAT custom branch unless we have a good reason, they already have a band-aid that's working fine. "Don't fix what ain't broken", if there is any issue left, we don't want to find out in the EUMETSAT production system

**Comment by jemacchi** [comment]
@Nuno Oliveira <nuno.oliveira@geo-solutions.it>
I'm trying to assign the issue in GWC  to myself and then move it in board,
but it seems that I need some permission to do it (
https://github.com/GeoWebCache/geowebcache/issues/880#)

Otherwise, I can comment solution in the ticket and current status (but not
to close the issue)

**Comment by simboss** [comment]
@jemacchi let me know once this sits onto EMSA branch so that we can deploy it and test it.

As far as the backport is concerned, yes, I would do that and we fin issues, we fix them and backport them.

**Comment by jemacchi** [comment]
@aaime  when possible please can you assist me on what @simboss  mentioned [here](https://github.com/geosolutions-it/eumetsat-EOWS/issues/361#issuecomment-697741205).
So far,  PR was merged on GWC master (https://github.com/GeoWebCache/geowebcache/commits/master ) and based on your comment [here](https://github.com/geosolutions-it/eumetsat-EOWS/issues/361#issuecomment-696679621)  would be to keep it out of EUMESAT.  

I will keep a note on this, considering EMSA, since not sure when it will be required to integrate on that project. CC: @nmco 
Let me know if I misunderstanding something (since we are talking about 2 different projects)

**Comment by simboss** [comment]
@jemacchi sorry my bad I wanted to say EUMETSAT.

Yes, we want to test the thing in DEV with @randomorder.
You need to cherry pick the change to a specific branch of our GWC fork which is documented [here](https://github.com/geosolutions-it/eumetsat-EOWS/wiki/Eumetsat-Pathfinder-2019-2020). Notice that on his branch as quick workaround we disabled entirely the code that manages parameters (see [this](https://github.com/geosolutions-it/geowebcache/commit/e53e9a28a0245fb4bce4fdec6fe2da0e3e10500d) commit) hence we need to *revert* this as well.

Is this clear now?

**Comment by jemacchi** [comment]
@simboss  I think I got it.
Basically, pick of my changes in GWC official and add them into [EOW-158](https://github.com/geosolutions-it/geowebcache/tree/EOWS-158) branch   (a cherry pick of them)  AND to consider the parameters thing (so we keep the workaround). Correct me if I'm wrong please.

**Comment by simboss** [comment]
Correct.

Give priority to MapCache work before.

**Comment by jemacchi** [comment]
@simboss @nmco 
I have checked steps with @aaime  about integration of this issue into EOW-158 branch.
I have my local branch ready for pushing, but I'm afraid that I have only permissions for reading, since I got this:

```
jmacchi@jmacchi:/prg/geosolutions/C068-EUMETSAT-CGI-2019-WP2/geowebcache/geowebcache$ git push geosolutions improve-gwc-metadata-access:improve-gwc-metadata-access
Username for 'https://github.com': jemacchi
Password for 'https://jemacchi@github.com': 
remote: Permission to geosolutions-it/geowebcache.git denied to jemacchi.
fatal: unable to access 'https://github.com/geosolutions-it/geowebcache/': The requested URL returned error: 403

```
If you enable writing permissions, then I can upload my branch, and create a PR for it.

**Comment by jemacchi** [comment]
@simboss latest [comment](https://github.com/geosolutions-it/eumetsat-EOWS/issues/361#issuecomment-704455071) was about permissions, so now I pushed code and created the PR as per you instructions @aaime 
@nmco  thanks for assistance on the permissions.

**Comment by nmco** [comment]
@jemacchi you shoudl request @aaime review not mine, @aaime can you have look at the backport?

**Comment by jemacchi** [comment]
@aaime this is the [PR](https://github.com/geosolutions-it/geowebcache/pull/6)

**Comment by aaime** [comment]
I still see this one in "review" column.. guess we can call it "done" right?

**Comment by simboss** [comment]
it's not done, we still have to merge it.
